dcrrates/rateserver/config.go

```diff
@@ -0,0 +1,184 @@
+	AltDNSNames       []string `long:"altdnsnames" description:"Specify additional dns names to use when generating the rpc server certificate" env:"DCRRATES_ALT_DNSNAMES" env-delim:","`
+}
+
+var defaultConfig = config{
+	AppDirectory:     dcrrates.DefaultAppDirectory,
+	GRPCListen:       defaultListen,
+	ExchangeCurrency: defaultBtcIndex,
+	ExchangeRefresh:  defaultExchangeRefresh,
+	ExchangeExpiry:   defaultExchangeExpiry,
+	AltDNSNames:      defaultAltDNSNames,
+}
+
+func loadConfig() (*config, error) {
+	cfg := defaultConfig
+
+	preCfg := cfg
+	preParser := flags.NewParser(&preCfg, flags.HelpFlag|flags.PassDoubleDash)
+	_, err := preParser.Parse()
+
+	if err != nil {
+		e, ok := err.(*flags.Error)
+		if !ok || e.Type != flags.ErrHelp {
+			preParser.WriteHelp(os.Stderr)
+		}
+		if ok && e.Type == flags.ErrHelp {
+			preParser.WriteHelp(os.Stdout)
+			os.Exit(0)
+		}
+		return nil, err
+	}
+
+	parser := flags.NewParser(&cfg, flags.Default)
+
+	if preCfg.AppDirectory != dcrrates.DefaultAppDirectory {
+		preCfg.AppDirectory = cleanAndExpandPath(preCfg.AppDirectory)
+	}
+
+	defaultConfigPath := filepath.Join(preCfg.AppDirectory, defaultConfigName)
+	if preCfg.ConfigPath == "" {
+		preCfg.ConfigPath = defaultConfigPath
+	} else {
+		preCfg.ConfigPath = cleanAndExpandPath(preCfg.ConfigPath)
+	}
+
+	_, err = os.Stat(preCfg.ConfigPath)
+	if os.IsNotExist(err) {
+		if preCfg.ConfigPath != defaultConfigPath {
+			fmt.Fprintln(os.Stderr, "No configuration file found at "+preCfg.ConfigPath)
+			os.Exit(0)
+		}
+	} else {
+		err = flags.NewIniParser(parser).ParseFile(preCfg.ConfigPath)
+		if err != nil {
+			fmt.Fprintln(os.Stderr, err)
+			parser.WriteHelp(os.Stderr)
+			return nil, fmt.Errorf("Unable to parse configuration file.")
+		}
+	}
+
+	_, err = parser.Parse()
+	if err != nil {
+		if e, ok := err.(*flags.Error); !ok || e.Type != flags.ErrHelp {
+			parser.WriteHelp(os.Stderr)
+		}
+		return nil, fmt.Errorf("Error parsing command line arguments: %v", err)
+	}
+
+	if cfg.AppDirectory != preCfg.AppDirectory {
+		cfg.AppDirectory = cleanAndExpandPath(cfg.AppDirectory)
+		err = os.MkdirAll(cfg.AppDirectory, 0700)
+		if err != nil {
+			return nil, fmt.Errorf("Application directory error: %v", err)
+		}
+	}
+
+	if cfg.CertificatePath == "" {
+		cfg.CertificatePath = filepath.Join(cfg.AppDirectory, dcrrates.DefaultCertName)
+	} else {
+		cfg.CertificatePath = cleanAndExpandPath(cfg.CertificatePath)
+	}
+	if cfg.KeyPath == "" {
+		cfg.KeyPath = filepath.Join(cfg.AppDirectory, dcrrates.DefaultKeyName)
+	} else {
+		cfg.KeyPath = cleanAndExpandPath(cfg.KeyPath)
+	}
+	if cfg.LogPath == "" {
+		cfg.LogPath = filepath.Join(cfg.AppDirectory, defaultLogDirName)
+	} else {
+		cfg.LogPath = cleanAndExpandPath(cfg.LogPath)
+	}
+
+	return &cfg, nil
+}
+
+// cleanAndExpandPath expands environment variables and leading ~ in the passed
+// path, cleans the result, and returns it.
+func cleanAndExpandPath(path string) string {
```

Comment:
This was copied from the root `config.go`. I wasn't certain how to document / copyright.